sound/soc/sof/loader.c

```diff
@@ -358,6 +359,86 @@ static int snd_sof_fw_ext_man_parse(struct snd_sof_dev *sdev,
 	return ext_man_size;
 }
 
+static int snd_sof_fw_ext_man4_parse(struct snd_sof_dev *sdev, const struct firmware *fw)
+{
+	struct sof_ipc4_fw_modules *module_entry;
+	struct sof_man4_fw_binary_header *fw_header;
+	struct sof_ext_manifest4_hdr *hdr;
+	struct sof_man4_module_config *fm_config;
+	struct sof_man4_module_entry *fm_entry;
+	int fw_offset;
+	int i;
+
+	if (fw->size < sizeof(hdr)) {
+		dev_err(sdev->dev, "Invalid fw size\n");
+		return -EINVAL;
+	}
+
+	hdr = (struct sof_ext_manifest4_hdr *)fw->data;
+
+	fw_offset = hdr->len;
+
+	fw_header = (struct sof_man4_fw_binary_header *)(fw->data + fw_offset +
+							 SOF_MAN4_FW_HDR_OFFSET);
+	dev_dbg(sdev->dev, " fw %s: header length %x, module num %d", fw_header->name,
+		fw_header->len, fw_header->num_module_entries);
+
+	fm_entry = (struct sof_man4_module_entry *)((void *)fw_header + fw_header->len);
+	fm_config = (struct sof_man4_module_config *)(fm_entry + fw_header->num_module_entries);
+
+	sdev->fw_modules = devm_kmalloc_array(sdev->dev, fw_header->num_module_entries,
+					      sizeof(*module_entry), GFP_KERNEL);
+	if (!sdev->fw_modules)
+		return -ENOMEM;
+
+	sdev->fw_module_num = fw_header->num_module_entries;
+	module_entry = sdev->fw_modules;
+
+	for (i = 0; i < fw_header->num_module_entries; i++) {
+		int dw_count;
+
+		dev_dbg(sdev->dev, "module %s : UUID %pUL, ", fm_entry->name,
+			fm_entry->uuid);
+
+		memcpy(module_entry->uuid, fm_entry->uuid, UUID_SIZE);
+		memcpy(module_entry->name, fm_entry->name, MAX_MODULE_NAME_LEN);
+
+		if (fm_entry->cfg_count)
+			module_entry->bss_size = fm_config[fm_entry->cfg_offset].is_bytes;
+
+		memcpy(&module_entry->type, &fm_entry->type, sizeof(fm_entry->type));
+
+		/* bringup fw starts at zero */
```

Comment:
why is this comment relevant here?

Reply:
bringup fw is not a common module but it is defined module config.